tree/dataframe/test/dataframe_datasetspec.py

```diff
@@ -0,0 +1,93 @@
+
+        # reuse the code from the C++ unit tests to create some files
+        code = ''' {
+        auto dfWriter0 = ROOT::RDataFrame(5)
+                            .Define("x", [](ULong64_t e) { return e; }, {"rdfentry_"})
+                            .Define("y", [](ULong64_t e) { return e + 10; }, {"rdfentry_"})
+                            .Define("z", [](ULong64_t e) { return e + 100; }, {"rdfentry_"});
+        dfWriter0.Range(0, 2).Snapshot<ULong64_t>("subTree", "PYspecTestFile2.root", {"z"});
+        dfWriter0.Range(2, 4).Snapshot<ULong64_t>("subTree", "PYspecTestFile3.root", {"z"});
+        dfWriter0.Range(4, 5).Snapshot<ULong64_t>("subTree", "PYspecTestFile4.root", {"z"});
+        dfWriter0.Range(0, 2).Snapshot<ULong64_t>("subTree1", "PYspecTestFile5.root", {"z"});
+        dfWriter0.Range(2, 4).Snapshot<ULong64_t>("subTree2", "PYspecTestFile6.root", {"z"});
+        dfWriter0.Snapshot<ULong64_t>("anotherTree", "PYspecTestFile7.root", {"z"});
+        }'''
+        ROOT.gInterpreter.Calc(code)
+
+    def tearDown(self):
+        for i in range(2, 8):
+            ROOT.gSystem.Unlink("PYspecTestFile" + str(i) + ".root")
+
+    # a test involving all 3 constructors, all possible valid ranges, all possible friends
+    def test_General(self):
+        ranges = [REntryRange(), REntryRange(1, 4), REntryRange(2,4), REntryRange(100), REntryRange(1, 100),
+                  REntryRange(2, 2), REntryRange(7, 7)]
+        expectedRess = [[100, 101, 102, 103, 104], [101, 102, 103], [102, 103], [100, 101, 102, 103, 104],
+                        [101, 102, 103, 104], [], []]
+        specs = []
+        
+        for r in ranges:
+            specs.append([])
+            specs[-1].append(RDatasetSpec("anotherTree", "PYspecTestFile7.root", r))
+            specs[-1].append(RDatasetSpec("subTree", ["PYspecTestFile2.root",
+                                                 "PYspecTestFile3.root",
+                                                 "PYspecTestFile4.root"], r))
+            specs[-1].append(RDatasetSpec([("subTree1", "PYspecTestFile5.root"),
+                                      ("subTree2", "PYspecTestFile6.root"),
+                                      ("subTree", "PYspecTestFile4.root")], r))
+
+            # for each spec add all possible combination of friends
+            for s in specs[-1]:
+                s.AddFriend("anotherTree", "PYspecTestFile7.root", "friendTree")
+                s.AddFriend("subTree", ["PYspecTestFile2.root",
+                                        "PYspecTestFile3.root",
+                                        "PYspecTestFile4.root"], "friendChain1")
+                s.AddFriend([("subTree1", "PYspecTestFile5.root"),
+                             ("subTree2", "PYspecTestFile6.root"),
+                             ("subTree", "PYspecTestFile4.root")], "friendChainN")
+
+        for i in range(len(ranges)):
+            for s in specs[i]:
```

Comment:
Following from the reasoning above, here you are accesing the i-th element of the `specs` list, but that list only has one element, right? It is that element (which is another list) the one who contains all the RDatasetSpec objects. So I guess this loop is not doing what you want really

Reply:
Yes it's a bit too hidden imho, maybe it would be more understandable with a list comprehension. Plus now that you explained it better I believe you're only adding friend info once since you write `for spec in specs[-1]` just a few lines below

Reply:
I am doing that every time. See it is:
```py
        for r in ranges:
            specs.append([])
            specs[-1].append(...)
            specs[-1].append(...)
            specs[-1].append(...)
            for s in specs[-1]:
                s.AddFriend(...)
```

Reply:
Ok, thanks for the clarification

Reply:
I will add comments, list comprehension seems to make it worse